File: agenkit-cpp/examples/adapters/openai-streaming.cpp
/**
 * @file openai-streaming.cpp
 * @brief Example demonstrating OpenAI streaming support
 *
 * This example shows how to stream responses from OpenAI models in real-time,
 * displaying text as it arrives from the API.
 *
 * Build:
 *   mkdir -p build && cd build
 *   cmake .. && make openai-streaming
 *
 * Run:
 *   export OPENAI_API_KEY="your-key-here"
 *   ./openai-streaming
 */

#include "agenkit/adapters/openai_agent.hpp"
#include "agenkit/core/message.hpp"
#include <iostream>
#include <cstdlib>

using namespace agenkit;
using namespace agenkit::adapters;
using namespace agenkit::core;

int main() {
    // Load API key from environment
    const char* api_key_env = std::getenv("OPENAI_API_KEY");
    if (!api_key_env) {
        std::cerr << "Error: OPENAI_API_KEY environment variable not set\n";
        return 1;
    }

    // Configure OpenAI
    OpenAIConfig config;
    config.api_key = api_key_env;
    config.model = "gpt-4o-mini";
    config.max_tokens = 1024;
    config.temperature = 0.7;

    OpenAIAgent gpt(config);

    std::cout << "=== OpenAI Streaming Example ===\n\n";
    std::cout << "Asking GPT to count to 10...\n\n";
    std::cout << "Response (streaming):\n";
    std::cout << std::string(60, '-') << "\n";

    // Create message
    auto message = Message::with_text("user", "Count to 10, one number per line.");

    // Stream response
    std::string full_response;
    auto result = gpt.stream(std::move(message), [&](const std::string& chunk) {
        std::cout << chunk << std::flush;
        full_response += chunk;
        return true;  // Continue streaming
    });

    if (result.is_err()) {
        std::cerr << "\nError: " << result.unwrap_err().message() << "\n";
        return 1;
    }

    std::cout << "\n" << std::string(60, '-') << "\n";
    std::cout << "\nFull response length: " << full_response.length() << " characters\n";

    // Example 2: Story generation
    std::cout << "\n\n=== Story Generation Example ===\n\n";
    std::cout << "Asking GPT to write a short story...\n\n";
    std::cout << "Response (streaming):\n";
    std::cout << std::string(60, '-') << "\n";

    auto message2 = Message::with_text(
        "user",
        "Write a very short story (3 sentences) about a robot learning to paint."
    );

    full_response.clear();
    result = gpt.stream(std::move(message2), [&](const std::string& chunk) {
        std::cout << chunk << std::flush;
        full_response += chunk;
        return true;  // Continue streaming
    });

    if (result.is_err()) {
        std::cerr << "\nError: " << result.unwrap_err().message() << "\n";
        return 1;
    }

    std::cout << "\n" << std::string(60, '-') << "\n";
    std::cout << "\nStreaming complete!\n";
    std::cout << "Total characters received: " << full_response.length() << "\n";

    // Example 3: Early termination
    std::cout << "\n\n=== Early Termination Example ===\n\n";
    std::cout << "Streaming with early stop after 50 characters...\n\n";
    std::cout << "Response (streaming):\n";
    std::cout << std::string(60, '-') << "\n";

    auto message3 = Message::with_text(
        "user",
        "Write a long essay about the history of computing."
    );

    full_response.clear();
    result = gpt.stream(std::move(message3), [&](const std::string& chunk) {
        std::cout << chunk << std::flush;
        full_response += chunk;

        // Stop after 50 characters
        if (full_response.length() >= 50) {
            std::cout << "\n[STOPPED EARLY]";
            return false;  // Stop streaming
        }
        return true;  // Continue streaming
    });

    std::cout << "\n" << std::string(60, '-') << "\n";
    std::cout << "\nStopped after " << full_response.length() << " characters\n";

    std::cout << "\nAll examples complete!\n";

    return 0;
}

File: agenkit-rust/examples/openai_streaming.rs
///! Example demonstrating OpenAI GPT streaming support.
///!
///! This example shows how to stream responses from OpenAI models in real-time,
///! displaying text as it arrives from the API.
///!
///! Run with:
///! ```bash
///! export OPENAI_API_KEY="your-key-here"
///! cargo run --example openai_streaming --features native
///! ```

use agenkit::adapters::openai::{OpenAIAgent, OpenAIConfig};
use agenkit::core::Message;
use futures::stream::StreamExt;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Configure OpenAI
    let config = OpenAIConfig {
        api_key: std::env::var("OPENAI_API_KEY")
            .expect("OPENAI_API_KEY environment variable not set"),
        model: "gpt-4o-mini".to_string(),
        temperature: 1.0,
        max_tokens: 1024,
        ..Default::default()
    };

    let adapter = OpenAIAgent::new(config);

    println!("=== OpenAI GPT Streaming Example ===\n");
    println!("Asking GPT to count to 10...\n");
    println!("Response (streaming):");
    println!("{}", "-".repeat(60));

    // Create message
    let message = Message::with_text("user", "Count to 10, one number per line.");

    // Stream response
    let mut stream = adapter.stream(message).await;
    let mut full_response = String::new();

    while let Some(result) = stream.next().await {
        match result {
            Ok(chunk) => {
                if let Some(text) = chunk.content_as_str() {
                    print!("{}", text);
                    full_response.push_str(text);
                }
            }
            Err(e) => {
                eprintln!("\nError: {:?}", e);
                return Err(e.into());
            }
        }
    }

    println!("\n{}", "-".repeat(60));
    println!("\nFull response length: {} characters", full_response.len());

    // Example 2: Story generation
    println!("\n\n=== Story Generation Example ===\n");
    println!("Asking GPT to write a short story...\n");
    println!("Response (streaming):");
    println!("{}", "-".repeat(60));

    let message = Message::with_text(
        "user",
        "Write a very short story (3 sentences) about a robot learning to paint.",
    );

    let mut stream = adapter.stream(message).await;
    let mut full_response = String::new();

    while let Some(result) = stream.next().await {
        match result {
            Ok(chunk) => {
                if let Some(text) = chunk.content_as_str() {
                    print!("{}", text);
                    full_response.push_str(text);
                }
            }
            Err(e) => {
                eprintln!("\nError: {:?}", e);
                return Err(e.into());
            }
        }
    }

    println!("\n{}", "-".repeat(60));
    println!("\nStreaming complete!");
    println!("Total characters received: {}", full_response.len());

    Ok(())
}
